File: packages/cloud-adapters/secrets/baremetal_test.go
package secrets

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"

	vault "github.com/hashicorp/vault/api"
)

// fakeTransit is a hermetic in-process stand-in for Vault's Transit
// Secrets Engine. It implements only the verbs the adapter exercises
// (mint, encrypt, decrypt, rotate, deletion-config, delete) and the
// canonical `vault:v<version>:<base64-payload>` ciphertext shape.
type fakeTransit struct {
	mu        sync.Mutex
	mount     string
	keys      map[string]int // name → current version
	allowDel  map[string]bool
}

func newFakeTransit(mount string) *fakeTransit {
	return &fakeTransit{
		mount:    mount,
		keys:     map[string]int{},
		allowDel: map[string]bool{},
	}
}

// route trims the `/v1/<mount>/` prefix and dispatches on the rest.
func (f *fakeTransit) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	prefix := "/v1/" + f.mount + "/"
	if !strings.HasPrefix(r.URL.Path, prefix) {
		http.Error(w, "no route", http.StatusNotFound)
		return
	}
	sub := strings.TrimPrefix(r.URL.Path, prefix)
	parts := strings.Split(sub, "/")

	f.mu.Lock()
	defer f.mu.Unlock()

	switch {
	case len(parts) == 2 && parts[0] == "keys" && (r.Method == http.MethodPost || r.Method == http.MethodPut):
		// mint-or-update at /keys/<name>. vault-api's Logical().Write uses PUT.
		f.keys[parts[1]] = 1
		writeData(w, nil)

	case len(parts) == 2 && parts[0] == "keys" && r.Method == http.MethodDelete:
		// delete at /keys/<name> — requires deletion_allowed
		if !f.allowDel[parts[1]] {
			http.Error(w, `{"errors":["deletion is not allowed for this key"]}`, http.StatusBadRequest)
			return
		}
		delete(f.keys, parts[1])
		delete(f.allowDel, parts[1])
		w.WriteHeader(http.StatusNoContent)

	case len(parts) == 3 && parts[0] == "keys" && parts[2] == "rotate" && (r.Method == http.MethodPost || r.Method == http.MethodPut):
		// rotate at /keys/<name>/rotate
		if _, ok := f.keys[parts[1]]; !ok {
			http.Error(w, `{"errors":["key not found"]}`, http.StatusNotFound)
			return
		}
		f.keys[parts[1]]++
		writeData(w, nil)

	case len(parts) == 3 && parts[0] == "keys" && parts[2] == "config" && (r.Method == http.MethodPost || r.Method == http.MethodPut):
		// set config at /keys/<name>/config — supports `deletion_allowed`
		body := decodeBody(r)
		if v, _ := body["deletion_allowed"].(bool); v {
			f.allowDel[parts[1]] = true
		}
		writeData(w, nil)

	case len(parts) == 2 && parts[0] == "encrypt" && (r.Method == http.MethodPost || r.Method == http.MethodPut):
		// encrypt at /encrypt/<name> — body has `plaintext` (base64)
		ver, ok := f.keys[parts[1]]
		if !ok {
			http.Error(w, `{"errors":["encryption key not found"]}`, http.StatusNotFound)
			return
		}
		body := decodeBody(r)
		pt, _ := body["plaintext"].(string) // already base64-encoded by adapter
		ct := fmt.Sprintf("vault:v%d:%s", ver, pt)
		writeData(w, map[string]any{"ciphertext": ct})

	case len(parts) == 2 && parts[0] == "decrypt" && (r.Method == http.MethodPost || r.Method == http.MethodPut):
		// decrypt at /decrypt/<name> — body has `ciphertext`
		if _, ok := f.keys[parts[1]]; !ok {
			http.Error(w, `{"errors":["encryption key not found"]}`, http.StatusNotFound)
			return
		}
		body := decodeBody(r)
		ct, _ := body["ciphertext"].(string)
		// Format: vault:v<n>:<base64-payload>. Extract last segment.
		idx := strings.LastIndex(ct, ":")
		if idx < 0 || !strings.HasPrefix(ct, "vault:v") {
			http.Error(w, `{"errors":["invalid ciphertext"]}`, http.StatusBadRequest)
			return
		}
		writeData(w, map[string]any{"plaintext": ct[idx+1:]})

	default:
		http.Error(w, "unhandled route: "+sub, http.StatusNotImplemented)
	}
}

// writeData wraps a payload in Vault's response envelope.
func writeData(w http.ResponseWriter, data map[string]any) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]any{
		"request_id": "fake",
		"data":       data,
	})
}

func decodeBody(r *http.Request) map[string]any {
	var body map[string]any
	_ = json.NewDecoder(r.Body).Decode(&body)
	return body
}

// newFakeVault spawns the fake Transit HTTP server, returns a vault
// client pointed at it.
func newFakeVault(t *testing.T, mount string) (*vault.Client, func()) {
	t.Helper()
	fake := newFakeTransit(mount)
	server := httptest.NewServer(fake)
	cfg := vault.DefaultConfig()
	cfg.Address = server.URL
	client, err := vault.NewClient(cfg)
	if err != nil {
		server.Close()
		t.Fatalf("vault client: %v", err)
	}
	client.SetToken("fake-token")
	return client, server.Close
}

func TestBaremetalAdapter_TransitRoundTrip(t *testing.T) {
	client, cleanup := newFakeVault(t, "transit")
	defer cleanup()
	ctx := context.Background()
	adapter := NewBaremetalAdapter(client, "transit")

	const keyName = "tenant-acme-pii"
	plaintext := []byte("sensitive-email@example.com")

	// Mint, encrypt, decrypt round-trip.
	if err := adapter.MintKey(ctx, keyName); err != nil {
		t.Fatalf("MintKey: %v", err)
	}
	ct, err := adapter.Encrypt(ctx, keyName, plaintext)
	if err != nil {
		t.Fatalf("Encrypt: %v", err)
	}
	if !strings.HasPrefix(ct, "vault:v1:") {
		t.Errorf("ciphertext shape: got %q, want vault:v1:<...>", ct)
	}
	pt, err := adapter.Decrypt(ctx, keyName, ct)
	if err != nil {
		t.Fatalf("Decrypt: %v", err)
	}
	if string(pt) != string(plaintext) {
		t.Errorf("round-trip: got %q want %q", pt, plaintext)
	}

	// Rotate: ciphertext from rotated key uses new version.
	if err := adapter.RotateKey(ctx, keyName); err != nil {
		t.Fatalf("RotateKey: %v", err)
	}
	ct2, err := adapter.Encrypt(ctx, keyName, plaintext)
	if err != nil {
		t.Fatalf("Encrypt after rotate: %v", err)
	}
	if !strings.HasPrefix(ct2, "vault:v2:") {
		t.Errorf("post-rotate ciphertext: got %q, want vault:v2:<...>", ct2)
	}
}

func TestBaremetalAdapter_CryptoShred(t *testing.T) {
	client, cleanup := newFakeVault(t, "transit")
	defer cleanup()
	ctx := context.Background()
	adapter := NewBaremetalAdapter(client, "transit")
	const keyName = "tenant-beta-pii"

	if err := adapter.MintKey(ctx, keyName); err != nil {
		t.Fatalf("MintKey: %v", err)
	}
	// Lock the key into existence: encrypt-decrypt works pre-shred.
	if _, err := adapter.Encrypt(ctx, keyName, []byte("data")); err != nil {
		t.Fatalf("Encrypt pre-shred: %v", err)
	}

	// Crypto-shred: deleteKey flips deletion_allowed + deletes.
	if err := adapter.DeleteKey(ctx, keyName); err != nil {
		t.Fatalf("DeleteKey (shred): %v", err)
	}

	// Subsequent encrypt with the same name MUST fail — the key is
	// gone and the operator should get an error rather than a silent
	// fall-through to mint-and-encrypt (Discussion #76 crypto-shred
	// guarantee: a deleted key cannot be impersonated by a fresh mint
	// with the same name unless the operator explicitly mints again).
	if _, err := adapter.Encrypt(ctx, keyName, []byte("data")); err == nil {
		t.Errorf("Encrypt after crypto-shred: expected error, got nil")
	}
}

// Sanity: an empty base64-encoded plaintext round-trips cleanly.
func TestBaremetalAdapter_EmptyPlaintext(t *testing.T) {
	client, cleanup := newFakeVault(t, "transit")
	defer cleanup()
	ctx := context.Background()
	adapter := NewBaremetalAdapter(client, "transit")
	const keyName = "edge"
	if err := adapter.MintKey(ctx, keyName); err != nil {
		t.Fatal(err)
	}
	ct, err := adapter.Encrypt(ctx, keyName, nil)
	if err != nil {
		t.Fatalf("Encrypt empty: %v", err)
	}
	pt, err := adapter.Decrypt(ctx, keyName, ct)
	if err != nil {
		t.Fatalf("Decrypt empty: %v", err)
	}
	// base64("") = "" — the round-trip returns an empty slice.
	if len(pt) != 0 {
		t.Errorf("empty plaintext round-trip: got %v, want empty", pt)
	}
	// Belt-and-suspenders: a sentinel value resolves correctly.
	_ = base64.StdEncoding
}

File: packages/cloud-adapters/secrets/baremetal.test.ts
/**
 * @jest-environment node
 *
 * Runtime contract test for the bare-metal Vault Transit SecretsAdapter.
 * Hermetic via a stubbed fetch — the test verifies wiring (each method
 * builds the right URL, method, body shape) rather than running an
 * actual Vault round-trip. The Go contract test (//packages/cloud-
 * adapters/secrets:secrets_test) does the round-trip via an in-process
 * Vault Transit HTTP fake.
 */
import { describe, beforeEach, it, expect } from "@jest/globals";

import { newBaremetalAdapter } from "./baremetal";

type FetchCall = { url: string; init: { method?: string; body?: string; headers?: Record<string, string> } };

interface FakeFetch {
  (url: string, init?: unknown): Promise<unknown>;
  calls: FetchCall[];
}

const newFakeFetch = (responder: () => { ok?: boolean; status?: number; _body?: unknown }): FakeFetch => {
  const calls: FetchCall[] = [];
  const fn = ((url: string, init: { method?: string; body?: string; headers?: Record<string, string> } = {}) => {
    calls.push({ url, init });
    const r = responder();
    return Promise.resolve({
      ok: r.ok ?? true,
      status: r.status ?? 200,
      json: () => Promise.resolve(r._body ?? {}),
      text: () =>
        Promise.resolve(typeof r._body === "string" ? r._body : JSON.stringify(r._body ?? {})),
    });
  }) as FakeFetch;
  fn.calls = calls;
  return fn;
};

const baseCfg = {
  endpoint: "http://vault.test:8200",
  token: "fake-token",
  mount: "transit",
};

describe("cloud-adapters/secrets/baremetal — runtime wiring contract", () => {
  let f: FakeFetch;

  beforeEach(() => {
    f = newFakeFetch(() => ({ _body: { data: {} } }));
  });

  it("mintKey POSTs to /v1/<mount>/keys/<name> with type aes256-gcm96", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.mintKey("tenant-acme");
    expect(f.calls).toHaveLength(1);
    expect(f.calls[0]!.url).toBe("http://vault.test:8200/v1/transit/keys/tenant-acme");
    expect(f.calls[0]!.init.method).toBe("POST");
    expect(JSON.parse(f.calls[0]!.init.body as string)).toEqual({ type: "aes256-gcm96" });
    expect(f.calls[0]!.init.headers!["X-Vault-Token"]).toBe("fake-token");
  });

  it("encrypt base64-encodes plaintext and returns ciphertext", async () => {
    f = newFakeFetch(() => ({ _body: { data: { ciphertext: "vault:v1:abc" } } }));
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    const ct = await adapter.encrypt("k1", new TextEncoder().encode("hello"));
    expect(ct).toBe("vault:v1:abc");
    expect(f.calls[0]!.url).toBe("http://vault.test:8200/v1/transit/encrypt/k1");
    const body = JSON.parse(f.calls[0]!.init.body as string);
    expect(body.plaintext).toBe(Buffer.from("hello").toString("base64"));
  });

  it("decrypt base64-decodes plaintext from the response", async () => {
    const b64 = Buffer.from("hello").toString("base64");
    f = newFakeFetch(() => ({ _body: { data: { plaintext: b64 } } }));
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    const pt = await adapter.decrypt("k1", "vault:v1:whatever");
    expect(new TextDecoder().decode(pt)).toBe("hello");
    expect(f.calls[0]!.url).toBe("http://vault.test:8200/v1/transit/decrypt/k1");
    expect(JSON.parse(f.calls[0]!.init.body as string)).toEqual({ ciphertext: "vault:v1:whatever" });
  });

  it("rotateKey POSTs to /v1/<mount>/keys/<name>/rotate", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.rotateKey("k1");
    expect(f.calls[0]!.url).toBe("http://vault.test:8200/v1/transit/keys/k1/rotate");
    expect(f.calls[0]!.init.method).toBe("POST");
  });

  it("deleteKey flips deletion_allowed then DELETEs (crypto-shred semantics)", async () => {
    f = newFakeFetch(() => ({ _body: { data: {} } }));
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.deleteKey("k1");
    expect(f.calls).toHaveLength(2);
    expect(f.calls[0]!.url).toBe("http://vault.test:8200/v1/transit/keys/k1/config");
    expect(JSON.parse(f.calls[0]!.init.body as string)).toEqual({ deletion_allowed: true });
    expect(f.calls[1]!.url).toBe("http://vault.test:8200/v1/transit/keys/k1");
    expect(f.calls[1]!.init.method).toBe("DELETE");
  });

  it("non-2xx responses surface as errors", async () => {
    f = newFakeFetch(() => ({ ok: false, status: 403, _body: '{"errors":["permission denied"]}' }));
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await expect(adapter.mintKey("k1")).rejects.toThrow(/403/);
  });
});
